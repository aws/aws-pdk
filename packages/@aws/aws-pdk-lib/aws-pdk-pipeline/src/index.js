"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
__exportStar(require("./pdk-pipeline"), exports);
__exportStar(require("./pdk-pipeline-ts-project"), exports);
__exportStar(require("./pdk-pipeline-py-project"), exports);
__exportStar(require("./pdk-pipeline-java-project"), exports);
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyJpbmRleC50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOzs7Ozs7Ozs7Ozs7QUFBQSxpREFBK0I7QUFDL0IsNERBQTBDO0FBQzFDLDREQUEwQztBQUMxQyw4REFBNEMiLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgKiBmcm9tIFwiLi9wZGstcGlwZWxpbmVcIjtcbmV4cG9ydCAqIGZyb20gXCIuL3Bkay1waXBlbGluZS10cy1wcm9qZWN0XCI7XG5leHBvcnQgKiBmcm9tIFwiLi9wZGstcGlwZWxpbmUtcHktcHJvamVjdFwiO1xuZXhwb3J0ICogZnJvbSBcIi4vcGRrLXBpcGVsaW5lLWphdmEtcHJvamVjdFwiO1xuIl19