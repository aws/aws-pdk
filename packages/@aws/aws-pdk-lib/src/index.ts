export * as construct from "./construct";
export * as project from "./project";