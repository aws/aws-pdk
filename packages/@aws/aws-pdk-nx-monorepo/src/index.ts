export * as project from "./project";
