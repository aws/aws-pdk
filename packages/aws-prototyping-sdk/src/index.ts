// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export class Hello {
  public sayHello() {
    return "hello, world!";
  }
}
