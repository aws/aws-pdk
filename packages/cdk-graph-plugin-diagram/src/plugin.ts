/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import {
  CdkGraph,
  CdkGraphArtifact,
  CdkGraphContext,
  ICdkGraphPlugin,
  IGraphPluginBindCallback,
  IGraphReportCallback,
  performGraphFilterPlan,
} from "@aws/cdk-graph";
import * as fs from "fs-extra";
import { toStream } from "ts-graphviz/adapter";
import {
  CONFIG_DEFAULTS,
  DEFAULT_DIAGRAM,
  DEFAULT_DIAGRAM_NAME,
  DiagramFormat,
  IDiagramConfig,
  IPluginConfig,
} from "./config";
import { IS_DEBUG } from "./internal/debug";
import { buildDiagram } from "./internal/graphviz/diagram";
import { resolveSvgAwsArchAssetImagesInline } from "./internal/utils/svg";

/**
 * CdkGraphDiagramPlugin is a {@link ICdkGraphPlugin CdkGraph Plugin} implementation for generating
 * diagram artifacts from the {@link CdkGraph} framework.
 */
export class CdkGraphDiagramPlugin implements ICdkGraphPlugin {
  /** Namespace for artifacts of the diagram plugin */
  static readonly ARTIFACT_NS = "DIAGRAM";
  /** Fixed id of the diagram plugin */
  static readonly ID = "diagram";
  /** Current semantic version of the diagram plugin */
  static readonly VERSION = "0.0.0";

  /** Get standardized artifact id for diagram artifacts */
  static artifactId(name: string, format: DiagramFormat): string {
    if (name === DEFAULT_DIAGRAM_NAME) {
      return `${this.ARTIFACT_NS}_${format.toUpperCase()}`;
    }
    return `${this.ARTIFACT_NS}_${name.toUpperCase()}_${format.toUpperCase()}`;
  }
  /** Get standardized artifact file name for diagram artifacts */
  static artifactFilename(name: string, format: DiagramFormat): string {
    if (name === DEFAULT_DIAGRAM_NAME) {
      return `${this.ARTIFACT_NS.toLowerCase()}.${format}`;
    }
    return `${this.ARTIFACT_NS.toLowerCase()}.${name}.${format}`;
  }

  /** @inheritdoc */
  get id(): string {
    return CdkGraphDiagramPlugin.ID;
  }
  /** @inheritdoc */
  get version(): string {
    return CdkGraphDiagramPlugin.VERSION;
  }

  /** @inheritdoc */
  readonly dependencies?: string[] = [];

  /** @internal */
  private _graph?: CdkGraph;

  /** @internal */
  private _config?: IPluginConfig;

  /** Get default dot artifact */
  get defaultDotArtifact(): CdkGraphArtifact | undefined {
    try {
      return this.getDiagramArtifact(DEFAULT_DIAGRAM_NAME, DiagramFormat.DOT);
    } catch {
      return;
    }
  }

  /** Get default PNG artifact */
  get defaultPngArtifact(): CdkGraphArtifact | undefined {
    try {
      return this.getDiagramArtifact(DEFAULT_DIAGRAM_NAME, DiagramFormat.PNG);
    } catch {
      return;
    }
  }

  /** Get diagram plugin config */
  get config(): IPluginConfig {
    if (this._config != null) {
      return this._config;
    }
    throw new Error(
      "Plugin config has not been instantiated, ensure bind is called"
    );
  }

  constructor(config?: IPluginConfig) {
    this._config = config;
  }

  /** Get diagram artifact for a given name and format */
  getDiagramArtifact(
    name: string,
    format: DiagramFormat
  ): CdkGraphArtifact | undefined {
    return this._graph?.graphContext?.getArtifact(
      CdkGraphDiagramPlugin.artifactId(name, format)
    );
  }

  /** @inheritdoc */
  bind: IGraphPluginBindCallback = (graph: CdkGraph): void => {
    this._graph = graph;
    const rc = (graph.config[CdkGraphDiagramPlugin.ID] ||
      {}) as Partial<IPluginConfig>;
    let diagrams: IDiagramConfig[] = [
      ...(rc.diagrams || []),
      ...(this._config?.diagrams || []),
    ];
    if (diagrams.length === 0) {
      diagrams = [DEFAULT_DIAGRAM];
    }
    this._config = {
      defaults: {
        ...CONFIG_DEFAULTS,
        ...rc.defaults,
        ...this._config?.defaults,
      },
      diagrams,
    };
  };

  private streamToBuffer = (stream: NodeJS.ReadableStream): Promise<Buffer> => {
    const chunks: Buffer[] = [];
    return new Promise((resolve, reject) => {
      stream.on("data", (chunk) => chunks.push(Buffer.from(chunk)));
      stream.on("error", (err) => reject(err));
      stream.on("end", () => resolve(Buffer.concat(chunks)));
    });
  };

  /** @inheritdoc */
  report?: IGraphReportCallback = async (
    context: CdkGraphContext
  ): Promise<void> => {
    const pluginConfig = this.config as Required<IPluginConfig>;

    for (const diagramConfig of pluginConfig.diagrams) {
      const config: IDiagramConfig = {
        ...(diagramConfig.ignoreDefaults ? {} : pluginConfig.defaults),
        ...diagramConfig,
      };

      let formats: DiagramFormat[] = Array.isArray(config.format)
        ? config.format
        : [config.format || DiagramFormat.PNG];
      if (!formats.length) {
        throw new Error(
          `Diagram config specifies empty list of formats; must provide at least 1 or undefined to use default.`
        );
      }

      // each diagram is destructive, so we need a clone
      const store = context.store.clone();

      if (config.filterPlan) {
        performGraphFilterPlan(store, config.filterPlan);
      }

      const generatePng = formats.includes(DiagramFormat.PNG);
      const generateSvg = generatePng || formats.includes(DiagramFormat.SVG);
      const generateDot = generateSvg || formats.includes(DiagramFormat.DOT);

      IS_DEBUG &&
        context.writeArtifact(
          this,
          "filtered.graph." + config.name + "",
          "debug/filtered-graph/" + config.name + ".json",
          JSON.stringify(store.serialize(), null, 2),
          "DEBUG"
        );

      const diagram = buildDiagram(store, {
        title: config.title,
        preset: config.filterPlan?.preset,
        theme: config.theme,
      });

      const dot = diagram.toDot();

      if (generateDot) {
        // Graphviz provider
        context.writeArtifact(
          this,
          CdkGraphDiagramPlugin.artifactId(config.name, DiagramFormat.DOT),
          CdkGraphDiagramPlugin.artifactFilename(
            config.name,
            DiagramFormat.DOT
          ),
          dot,
          `Diagram generated "dot" file for ${config.name} - "${config.title}"`
        );

        if (generateSvg) {
          const svgFile = path.join(
            context.outdir,
            CdkGraphDiagramPlugin.artifactFilename(
              config.name,
              DiagramFormat.SVG
            )
          );

          const svg = await this.streamToBuffer(
            await toStream(dot, { format: "svg" })
          );
          const resolvedSvg = await resolveSvgAwsArchAssetImagesInline(
            svg.toString()
          );

          fs.ensureDirSync(path.dirname(svgFile));
          fs.writeFileSync(svgFile, resolvedSvg);

          context.logArtifact(
            this,
            CdkGraphDiagramPlugin.artifactId(config.name, DiagramFormat.SVG),
            svgFile,
            `Diagram generated "svg" file for ${config.name} - "${config.title}"`
          );

          if (generatePng) {
            const pngFile = path.join(
              context.outdir,
              CdkGraphDiagramPlugin.artifactFilename(
                config.name,
                DiagramFormat.PNG
              )
            );

            try {
              const png = await this.streamToBuffer(
                await toStream(dot, {
                  format: "png",
                })
              );

              fs.ensureDirSync(path.dirname(pngFile));
              fs.writeFileSync(pngFile, png);

              context.logArtifact(
                this,
                CdkGraphDiagramPlugin.artifactId(
                  config.name,
                  DiagramFormat.PNG
                ),
                pngFile,
                `Diagram generated "png" file for ${config.name} - "${config.title}"`
              );
            } catch (error) {
              console.error(error);
              throw new Error(
                `Failed to generate PNG diagram for ${
                  config.name
                } at "${pngFile}" - ${String(error)}`
              );
            }
          }
        }
      }

      // NB: add drawio provider support here
    }
  };
}
