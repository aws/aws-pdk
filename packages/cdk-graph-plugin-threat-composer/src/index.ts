/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./plugin";
export * from "./model-generator/types";
