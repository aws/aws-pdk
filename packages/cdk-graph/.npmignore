# ~~ Generated by projen. To modify, edit .projenrc.js and run "pnpm exec projen".
/.projen/
/test-reports/
junit.xml
/coverage/
/.prettierignore
/.prettierrc.json
/test/
/tsconfig.dev.json
/src/
!/lib/
!/lib/**/*.js
!/lib/**/*.d.ts
dist
/tsconfig.json
/.github/
/.vscode/
/.idea/
/.projenrc.js
tsconfig.tsbuildinfo
/.eslintrc.json
!.jsii
/.gitattributes
