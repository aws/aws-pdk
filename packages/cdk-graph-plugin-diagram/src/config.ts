/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { FilterPreset, IGraphFilterPlan } from "@aws/cdk-graph";
import { GraphThemeConfigProp } from "./internal/graphviz";

/**
 * Supported diagram formats that can be generated.
 *
 * Extended formats are automatically generated, for example if you generate "png" which extends
 * "svg" which extends "dot", the resulting generated files will be all aforementioned.
 */
export enum DiagramFormat {
  /**
   * Graphviz [DOT Language](https://graphviz.org/doc/info/lang.html)
   *
   * @default
   */
  DOT = "dot",
  /**
   * [SVG](https://developer.mozilla.org/en-US/docs/Web/SVG) generated
   * using [dot-wasm](https://hpcc-systems.github.io/hpcc-js-wasm/classes/graphviz.Graphviz.html)
   * from {@link DiagramFormat.DOT} file.
   *
   * @extends {DiagramFormat.DOT}
   */
  SVG = "svg",
  /**
   * [PNG](https://en.wikipedia.org/wiki/Portable_Network_Graphics) generated
   * using [sharp](https://sharp.pixelplumbing.com/api-output#png) from {@link DiagramFormat.SVG} file
   *
   * @extends {DiagramFormat.SVG}
   */
  PNG = "png",

  // TODO: add drawio support
}

/**
 * Positional coordinates for a node in inches
 */
export interface INodePosition {
  readonly x: number;
  readonly y: number;
}

/**
 * Base config to specific a unique diagram to be generated.
 * @struct
 */
export interface IDiagramConfigBase {
  /**
   * The output format(s) to generated.
   * @default `DiagramFormat.PNG` - which will through extension also generate `DiagramFormat.SVG` and `DiagramFormat.DOT`
   */
  readonly format?: DiagramFormat | DiagramFormat[];

  /** Graph {@link IGraphFilterPlan Filter Plan}  used to generate a unique diagram */
  readonly filterPlan?: IGraphFilterPlan;

  /** Config for graph theme */
  readonly theme?: GraphThemeConfigProp;

  /** Config for predetermined node positions given their CDK construct id. */
  readonly nodePositions?: { [cdkConstructId: string]: INodePosition };
}

/** Diagram configuration definition
 * @struct
 */
export interface IDiagramConfig extends IDiagramConfigBase {
  /**
   * Name of the diagram. Used as the basename of the generated file(s) which gets the extension appended.
   */
  readonly name: string;

  /**
   * The title of the diagram.
   */
  readonly title: string;

  /**
   * Indicates if default diagram config is applied as defaults to this config.
   *
   * @default false
   */
  readonly ignoreDefaults?: boolean;
}

/**
 * Plugin configuration for diagram plugin.
 * @struct
 */
export interface IPluginConfig {
  /** Default configuration to apply to all diagrams */
  readonly defaults?: IDiagramConfigBase;
  /** List of diagram configurations to generate diagrams */
  readonly diagrams?: IDiagramConfig[];
}

/** Default diagram title */
export const DEFAULT_DIAGRAM_TITLE = "Cloud Diagram";

/** Default diagram name */
export const DEFAULT_DIAGRAM_NAME = "diagram";

/** Default configuration settings */
export const CONFIG_DEFAULTS: IDiagramConfigBase = {
  format: [DiagramFormat.PNG],
  filterPlan: {
    preset: FilterPreset.COMPACT,
  },
};

/** Default diagram config */
export const DEFAULT_DIAGRAM: IDiagramConfig = {
  name: DEFAULT_DIAGRAM_NAME,
  title: DEFAULT_DIAGRAM_TITLE,
};
