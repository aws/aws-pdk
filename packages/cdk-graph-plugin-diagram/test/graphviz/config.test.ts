/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { CdkGraph, FilterPreset, getConstructUUID } from "@aws-pdk/cdk-graph";
import { FixtureApp } from "@aws-pdk/cdk-graph/test/__fixtures__/apps";
import * as fs from "fs-extra";
import * as testUtils from "./test-utils";
import { CdkGraphDiagramPlugin, DiagramFormat } from "../../src";

jest.setTimeout(90000); // CI tests timeout occasionally so increase to large timeout buffer

const makeCdkOutdir = async (name: string) =>
  testUtils.makeCdkOutDir("config", name);

describe("config", () => {
  describe("default", () => {
    let outdir: string;
    let app: FixtureApp;
    let graph: CdkGraph;
    let plugin: CdkGraphDiagramPlugin;

    beforeAll(async () => {
      outdir = await makeCdkOutdir("default");

      app = new FixtureApp({ outdir });
      plugin = new CdkGraphDiagramPlugin();
      graph = new CdkGraph(app, {
        plugins: [plugin],
      });
      app.synth();
      await graph.report();
    });

    it("should generate default diagram", async () => {
      expect(plugin.defaultDotArtifact).toBeDefined();
      expect(
        await fs.pathExists(plugin.defaultDotArtifact!.filepath)
      ).toBeTruthy();
      expect(
        testUtils.cleanseDotSnapshot(
          await fs.readFile(plugin.defaultDotArtifact!.filepath, {
            encoding: "utf-8",
          })
        )
      ).toMatchSnapshot();

      await testUtils.expectToMatchImageSnapshot(
        plugin.defaultPngArtifact!.filepath,
        "default"
      );
    });
  });
  describe("plan", () => {
    let outdir: string;
    let app: FixtureApp;
    let graph: CdkGraph;
    let plugin: CdkGraphDiagramPlugin;

    beforeAll(async () => {
      outdir = await makeCdkOutdir("default");

      app = new FixtureApp({ outdir });
      plugin = new CdkGraphDiagramPlugin({
        diagrams: [
          {
            name: "compact",
            title: "Compact Diagram",
            filterPlan: {
              preset: FilterPreset.COMPACT,
            },
          },
          {
            name: "verbose",
            title: "Verbose Diagram",
            format: DiagramFormat.PNG,
            ignoreDefaults: true,
          },
          {
            name: "focus",
            title: "Focus Data Diagram (non-extraneous)",
            filterPlan: {
              focus: (store) =>
                store.getNode(getConstructUUID(app.stack.dataLayer)),
              preset: FilterPreset.NON_EXTRANEOUS,
            },
            ignoreDefaults: true,
          },
          {
            name: "focus-nohoist",
            title: "Focus Website Diagram (noHoist, verbose)",
            filterPlan: {
              focus: {
                node: (store) =>
                  store.getNode(getConstructUUID(app.stack.website)),
                noHoist: true,
              },
            },
            ignoreDefaults: true,
          },
        ],
      });
      graph = new CdkGraph(app, {
        plugins: [plugin],
      });
      app.synth();
      await graph.report();
    });

    it("should not generate default diagram", async () => {
      expect(plugin.defaultDotArtifact).not.toBeDefined();
    });

    it.each(["compact", "verbose", "focus", "focus-nohoist"])(
      "%s",
      async (diagramName) => {
        const artifact = plugin.getDiagramArtifact(
          diagramName,
          DiagramFormat.PNG
        );
        expect(artifact).toBeDefined();
        expect(await fs.pathExists(artifact!.filepath)).toBeTruthy();

        await testUtils.expectToMatchImageSnapshot(
          artifact!.filepath,
          diagramName,
          // focus and focus-nohoist rendering of fonts is a bit more variant on CI
          diagramName.startsWith("focus") ? 0.1 : undefined,
          diagramName.startsWith("focus") ? 0.2 : undefined
        );
      }
    );
  });
});
