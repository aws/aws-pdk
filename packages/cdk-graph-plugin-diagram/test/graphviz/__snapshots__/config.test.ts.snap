// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config default should generate default diagram 1`] = `
"digraph \\"ROOT\\" {
  label = \\"Cloud Diagram\\";
  imagepath = \\"<aws_arch_asset_directory\\";
  fontname = \\"Roboto Condensed\\";
  class = \\"font_regular\\";
  center = true;
  compound = true;
  concentrate = true;
  dpi = 300;
  fontcolor = \\"#000000\\";
  fontnames = \\"ps\\";
  fontsize = 14;
  forcelabels = true;
  labelloc = \\"tc\\";
  nodesep = 0.6;
  pad = \\"%2,%1\\";
  rankdir = \\"TB\\";
  ranksep = 0.75;
  ratio = \\"compress\\";
  remincross = true;
  size = \\"%1024,%1024!\\";
  splines = \\"ortho\\";
  stylesheet = \\"https://fonts.googleapis.com/css2?family=Roboto+Condensed:wght@300;400;700&display=swap\\";
  bgcolor = \\"#FFFFFF\\";
  edge [
    fontname = \\"Roboto Condensed\\";
    class = \\"font_light_italic\\";
    dir = \\"both\\";
    color = \\"#5A6B861A\\";
    penwidth = 0.75;
    arrowhead = \\"none\\";
    arrowtail = \\"normal\\";
    arrowsize = 0.5;
    fontsize = 9;
    style = \\"solid\\";
    fontcolor = \\"#5A6B861A\\";
  ];
  node [
    fontname = \\"Roboto Condensed\\";
    class = \\"font_regular\\";
    shape = \\"box\\";
    style = \\"solid\\";
    fixedsize = false;
    width = 0.25;
    height = 0.25;
    labelloc = \\"c\\";
    imagescale = true;
    fontsize = 11;
    penwidth = 0.25;
    fontcolor = \\"#232F3E\\";
  ];
  subgraph \\"cluster_stack_FixtureStack\\" {
    fontname = \\"Roboto Condensed\\";
    class = \\"font_light\\";
    labelloc = \\"tc\\";
    fontsize = 12;
    style = \\"solid,bold,filled\\";
    label = \\"FixtureStack\\";
    fillcolor = \\"#5A6B861A\\";
    margin = 10;
    color = \\"#000000CC\\";
    bgcolor = \\"#5A6B861A\\";
    pencolor = \\"transparent\\";
    fontcolor = \\"#000000CC\\";
    subgraph \\"cluster_FixtureStackNetworkingLayerE66F4752\\" {
      fontname = \\"Roboto Condensed\\";
      class = \\"font_light\\";
      labelloc = \\"tc\\";
      fontsize = 12;
      style = \\"rounded,dashed\\";
      label = \\"NetworkingLayer\\";
      \\"node_FixtureStackNetworkingLayerWebSecurityGroup843D70DB\\" [
        label = \\"WebSecurity\\\\nGroup\\";
        comment = \\"nodeType:RESOURCE(AWS::EC2::SecurityGroup)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.46;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"compute/ec2/service_icon.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
      \\"node_FixtureStackNetworkingLayerwebserverrole4C3EB3B3\\" [
        label = \\"WebserverRole\\";
        comment = \\"nodeType:RESOURCE(AWS::IAM::Role)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.23;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"security_identity_compliance/iam/role.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
      \\"node_FixtureStackNetworkingLayerWebServer589378AD\\" [
        label = \\"WebServer\\";
        comment = \\"nodeType:RESOURCE(AWS::EC2::Instance)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.23;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"compute/ec2/instance/c5.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
      \\"node_FixtureStackNetworkingLayerBastion092C2DFE\\" [
        label = \\"Bastion\\";
        comment = \\"nodeType:RESOURCE(AWS::EC2::Instance)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.23;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"compute/ec2/instance/t2.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
    }
    subgraph \\"cluster_FixtureStackDataLayer450628AB\\" {
      fontname = \\"Roboto Condensed\\";
      class = \\"font_light\\";
      labelloc = \\"tc\\";
      fontsize = 12;
      style = \\"rounded,dashed\\";
      label = \\"DataLayer\\";
      \\"node_FixtureStackDataLayerReadRoleEE6BC0A6\\" [
        label = \\"ReadRole\\";
        comment = \\"nodeType:RESOURCE(AWS::IAM::Role)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.23;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"security_identity_compliance/iam/role.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
    }
    subgraph \\"cluster_FixtureStackWebsite0ABF8522\\" {
      fontname = \\"Roboto Condensed\\";
      class = \\"font_light\\";
      labelloc = \\"tc\\";
      fontsize = 12;
      style = \\"rounded,dashed\\";
      label = \\"Website\\";
      \\"node_FixtureStackWebsiteWebDistroA236B451\\" [
        label = \\"WebDistro\\";
        comment = \\"nodeType:RESOURCE(AWS::CloudFront::Distribution)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.23;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"networking_content_delivery/cloudfront/service_icon.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
    }
    subgraph \\"cluster_FixtureStackEdgeCasesE05603B9\\" {
      fontname = \\"Roboto Condensed\\";
      class = \\"font_light\\";
      labelloc = \\"tc\\";
      fontsize = 12;
      style = \\"rounded,dashed\\";
      label = \\"EdgeCases\\";
      \\"node_FixtureStackEdgeCasesLambdaWithAlias3D0F36DD\\" [
        label = \\"LambdaWith\\\\nAlias\\";
        comment = \\"nodeType:RESOURCE(AWS::Lambda::Function)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.46;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"compute/lambda/lambda_function.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
      \\"node_FixtureStackEdgeCasesImportedBucketD2A18770\\" [
        label = \\"ImportedBucket\\";
        comment = \\"nodeType:CFN_RESOURCE(AWS::S3::Bucket)\\";
        shape = \\"box\\";
        style = \\"filled,dotted\\";
        fixedsize = true;
        width = 1;
        height = 1.23;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 1;
        fillcolor = \\"#5A6B861A\\";
        image = \\"storage/s3/bucket.svg\\";
        color = \\"#5A6B8633\\";
        fontcolor = \\"#5A6B86\\";
      ];
      \\"node_FixtureStackEdgeCasesImportedLambda2613949D\\" [
        label = \\"ImportedLambda\\";
        comment = \\"nodeType:CFN_RESOURCE(AWS::Lambda::Function)\\";
        shape = \\"box\\";
        style = \\"filled,dotted\\";
        fixedsize = true;
        width = 1;
        height = 1.23;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 1;
        fillcolor = \\"#5A6B861A\\";
        image = \\"compute/lambda/lambda_function.svg\\";
        color = \\"#5A6B8633\\";
        fontcolor = \\"#5A6B86\\";
      ];
      \\"node_FixtureStackEdgeCasesImportedResourceRoleAF826714\\" [
        label = \\"Imported\\\\nResourceRole\\";
        comment = \\"nodeType:RESOURCE(AWS::IAM::Role)\\";
        shape = \\"box\\";
        style = \\"solid,rounded\\";
        fixedsize = true;
        width = 1;
        height = 1.46;
        labelloc = \\"b\\";
        imagescale = true;
        imagepos = \\"tc\\";
        penwidth = 0;
        fillcolor = \\"transparent\\";
        image = \\"security_identity_compliance/iam/role.svg\\";
        color = \\"#232F3E\\";
        fontcolor = \\"#232F3E\\";
      ];
    }
  }
  \\"node_FixtureStackNetworkingLayerWebServer589378AD\\" -> \\"node_FixtureStackNetworkingLayerwebserverrole4C3EB3B3\\" [
    id = \\"REF:Z2qpAph\\";
    penwidth = 0.75;
    arrowhead = \\"normal\\";
    arrowtail = \\"none\\";
    arrowsize = 0.75;
    style = \\"solid\\";
    color = \\"#5A6B86\\";
    fontcolor = \\"#5A6B86\\";
  ];
  \\"node_FixtureStackNetworkingLayerWebServer589378AD\\" -> \\"node_FixtureStackNetworkingLayerWebSecurityGroup843D70DB\\" [
    id = \\"ATT:5q4po\\";
    penwidth = 0.75;
    arrowhead = \\"normal\\";
    arrowtail = \\"none\\";
    arrowsize = 0.75;
    style = \\"solid\\";
    color = \\"#5A6B86\\";
    fontcolor = \\"#5A6B86\\";
  ];
  \\"node_FixtureStackEdgeCasesImportedResourceRoleAF826714\\" -> \\"node_FixtureStackEdgeCasesImportedBucketD2A18770\\" [
    id = \\"IMP:14snQE\\";
    penwidth = 0.75;
    arrowhead = \\"normal\\";
    arrowtail = \\"none\\";
    arrowsize = 0.75;
    style = \\"solid\\";
    color = \\"#5A6B86\\";
    fontcolor = \\"#5A6B86\\";
  ];
  \\"node_FixtureStackEdgeCasesImportedResourceRoleAF826714\\" -> \\"node_FixtureStackEdgeCasesImportedLambda2613949D\\" [
    id = \\"IMP:1IosnQ\\";
    penwidth = 0.75;
    arrowhead = \\"normal\\";
    arrowtail = \\"none\\";
    arrowsize = 0.75;
    style = \\"solid\\";
    color = \\"#5A6B86\\";
    fontcolor = \\"#5A6B86\\";
  ];
}"
`;
